pkg/streamingpromql/operators/aggregation.go

```diff
@@ -136,14 +140,121 @@ func (a *Aggregation) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadat
 	return seriesMetadata, nil
 }
 
-func (a *Aggregation) labelsForGroup(m labels.Labels, lb *labels.Builder) labels.Labels {
+// seriesToGroupLabelsStringFunc returns a function that computes a string representation of the output group labels for the given input series.
+//
+// The returned function returns a byte slice rather than a string to make it possible to avoid unnecessarily allocating a string.
+func (a *Aggregation) seriesToGroupLabelsStringFunc() func(labels.Labels) []byte {
+	// Why not just use the labels.Labels computed by seriesToGroupLabelsFunc() and call String() on it?
+	//
+	// Most of the time, we don't need the labels.Labels instance, as we expect there are far fewer output groups than input series,
+	// and we only need the labels.Labels instance once per output group.
+	// However, we always need to compute the string representation for each input series, so we can look up its corresponding
+	// output group. And we can do this without allocating a string by returning just the bytes that make up the string.
+	// There's not much point in using the hash of the group labels as we always need the string (or the labels.Labels) to ensure
+	// there are no hash collisions - so we might as well just go straight to the string representation.
+	//
+	// Furthermore, labels.Labels.String() doesn't allow us to reuse the buffer used when producing the string or to return a byte slice,
+	// whereas this method does.
+	// This saves us allocating a new buffer and string for every single input series, which has a noticeable performance impact.
+
+	if a.Without {
+		b := make([]byte, 0, 1024)
+		buf := bytes.NewBuffer(b)
+
+		return func(l labels.Labels) []byte {
+			buf.Reset()
+			nextGroupingIndex := 0
+			outputLabelCount := 0
+
+			l.Range(func(l labels.Label) {
+				if l.Name == labels.MetricName {
+					return
+				}
+
+				for nextGroupingIndex < len(a.Grouping) && l.Name > a.Grouping[nextGroupingIndex] {
+					nextGroupingIndex++
+				}
+
+				if nextGroupingIndex == len(a.Grouping) || l.Name != a.Grouping[nextGroupingIndex] {
+					if outputLabelCount > 0 {
+						buf.WriteByte(',')
+						buf.WriteByte(' ')
+					}
+
+					buf.WriteString(l.Name)
+					buf.WriteByte('=')
+					buf.Write(strconv.AppendQuote(buf.AvailableBuffer(), l.Value))
+
+					outputLabelCount++
+				}
+			})
+
+			return buf.Bytes()
+		}
+	}
+
+	if len(a.Grouping) == 0 {
+		// We're grouping all input series into a single group with no labels.
+		var b []byte
+
+		return func(_ labels.Labels) []byte {
+			return b
+		}
+	}
+
+	b := make([]byte, 0, 1024)
+	buf := bytes.NewBuffer(b)
+	return func(l labels.Labels) []byte {
+		buf.Reset()
+		outputLabelCount := 0
+
+		for _, labelName := range a.Grouping {
+			labelValue := l.Get(labelName)
```

Comment:
It would require reaching into the labels somehow.

However, I've just noticed that there's a `BytesWithLabels` / `BytesWithoutLabels` method that largely does what I've implemented here that can do exactly what you suggest. I'll rework this to use that.

Reply:
I've reworked this to use `BytesWithLabels` / `BytesWithoutLabels` in a93c2d6.

Performance for our benchmark cases is largely unchanged, although we don't benchmark a scenario where we're grouping by multiple labels. Given how many benchmarks we already have, I'm hesitant to add more, but I'm confident your suggestion will have made things better for that scenario.

Reply:
I think it's pretty uncommon to have a significant amount of labels. I don't feel the need to benchmark this either.